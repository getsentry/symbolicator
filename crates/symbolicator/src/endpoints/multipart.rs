//! Helper utilities to handle HTTP multipart bodies.

use std::io::SeekFrom;

use axum::extract::multipart::Field;
use axum::http::StatusCode;
use futures::prelude::*;
use tokio::fs::File;
use tokio::io::AsyncSeekExt;
use tokio_util::io::StreamReader;

use super::ResponseError;

/// Stream a multipart body to a file.
///
/// The file's cursor will be set to the beginning of the file again.
pub async fn stream_multipart_file(field: Field<'_>, file: &mut File) -> Result<(), ResponseError> {
    let mut field_reader =
        StreamReader::new(field.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err)));
    tokio::io::copy(&mut field_reader, file).await?;
    file.seek(SeekFrom::Start(0)).await?;
    Ok(())
}

/// Read a multipart body into memory.
///
/// This respects a maximum size.
pub async fn read_multipart_data(
    mut field: Field<'_>,
    max_size: usize,
) -> Result<Vec<u8>, ResponseError> {
    let mut data = Vec::with_capacity(512);

    while let Some(chunk) = field.next().await {
        let chunk = chunk?;
        if (data.len() + chunk.len()) > max_size {
            let err = anyhow::anyhow!(
                "Field {} larger than {} bytes",
                field.name().unwrap_or_default(),
                max_size
            );
            return Err((StatusCode::PAYLOAD_TOO_LARGE, err).into());
        }
        data.extend_from_slice(&chunk);
    }
    Ok(data)
}
