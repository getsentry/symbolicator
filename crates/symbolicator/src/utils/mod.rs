pub mod addr;
pub mod compression;
pub mod futures;
pub mod gcs;
pub mod hex;
pub mod http;
pub mod paths;
pub mod sentry;
