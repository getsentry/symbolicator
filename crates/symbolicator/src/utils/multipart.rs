use actix_web::{dev::Payload, error, multipart, Error};
use futures::{compat::Stream01CompatExt, StreamExt};

use crate::sources::SourceConfig;
use crate::types::RequestOptions;

const MAX_JSON_SIZE: usize = 1_000_000;

pub async fn read_multipart_data(
    field: multipart::Field<Payload>,
    max_size: usize,
) -> Result<Vec<u8>, Error> {
    let mut body = Vec::with_capacity(512);
    let mut stream = field.compat();

    while let Some(chunk) = stream.next().await {
        let chunk = chunk?;

        if (body.len() + chunk.len()) > max_size {
            return Err(error::ErrorBadRequest("payload too large"));
        }

        body.extend_from_slice(&chunk);
    }

    Ok(body)
}

pub async fn read_multipart_file(field: multipart::Field<Payload>) -> Result<Vec<u8>, Error> {
    let mut body = Vec::with_capacity(512);
    let mut stream = field.compat();

    while let Some(chunk) = stream.next().await {
        body.extend_from_slice(&chunk?);
    }

    Ok(body)
}

pub async fn read_multipart_sources(
    field: multipart::Field<Payload>,
) -> Result<Vec<SourceConfig>, Error> {
    let data = read_multipart_data(field, MAX_JSON_SIZE).await?;
    Ok(serde_json::from_slice(&data)?)
}

pub async fn read_multipart_request_options(
    field: multipart::Field<Payload>,
) -> Result<RequestOptions, Error> {
    let data = read_multipart_data(field, MAX_JSON_SIZE).await?;
    Ok(serde_json::from_slice(&data)?)
}
