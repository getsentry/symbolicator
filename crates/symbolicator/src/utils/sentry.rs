use symbolicator_sources::ObjectId;

/// Write own data to [`sentry::Scope`], only the subset that is considered useful for debugging.
pub trait ConfigureScope {
    /// Writes information to the given scope.
    fn to_scope(&self, scope: &mut sentry::Scope);

    /// Configures the current scope.
    fn configure_scope(&self) {
        sentry::configure_scope(|scope| self.to_scope(scope));
    }
}

impl ConfigureScope for ObjectId {
    fn to_scope(&self, scope: &mut sentry::Scope) {
        scope.set_tag(
            "object_id.code_id",
            self.code_id
                .as_ref()
                .map(ToString::to_string)
                .unwrap_or_else(|| "None".to_string()),
        );
        scope.set_tag(
            "object_id.code_file_basename",
            self.code_file_basename().unwrap_or("None"),
        );
        scope.set_tag(
            "object_id.debug_id",
            self.debug_id
                .as_ref()
                .map(ToString::to_string)
                .unwrap_or_else(|| "None".to_string()),
        );
        scope.set_tag(
            "object_id.debug_file_basename",
            self.debug_file_basename().unwrap_or("None"),
        );
        scope.set_tag("object_id.object_type", self.object_type.to_string());
    }
}
