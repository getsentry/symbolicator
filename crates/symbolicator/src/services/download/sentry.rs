//! Support to download from sentry sources.
//!
//! Specifically this supports the [`SentrySourceConfig`] source, which allows
//! to fetch files which were directly uploaded to Sentry itself.

use std::fmt;
use std::path::Path;
use std::sync::Arc;
use std::time::{Duration, Instant};

use futures::TryStreamExt;
use parking_lot::Mutex;
use reqwest::{header, StatusCode};
use sentry::SentryFutureExt;
use serde::Deserialize;
use thiserror::Error;
use url::Url;

use super::{
    content_length_timeout, DownloadError, DownloadStatus, FileType, RemoteDif, RemoteDifUri,
    USER_AGENT,
};
use crate::config::Config;
use crate::sources::SentrySourceConfig;
use crate::types::ObjectId;
use crate::utils::futures::{self as future_utils, CancelOnDrop};

/// The Sentry-specific [`RemoteDif`].
#[derive(Debug, Clone)]
pub struct SentryRemoteDif {
    pub source: Arc<SentrySourceConfig>,
    pub file_id: SentryFileId,
}

impl From<SentryRemoteDif> for RemoteDif {
    fn from(source: SentryRemoteDif) -> Self {
        Self::Sentry(source)
    }
}

impl SentryRemoteDif {
    pub fn new(source: Arc<SentrySourceConfig>, file_id: SentryFileId) -> Self {
        Self { source, file_id }
    }

    pub fn uri(&self) -> RemoteDifUri {
        format!("sentry://project_debug_file/{}", self.file_id).into()
    }

    /// Returns the URL from which to download this object file.
    pub fn url(&self) -> Url {
        let mut url = self.source.url.clone();
        url.query_pairs_mut().append_pair("id", &self.file_id.0);
        url
    }
}

/// An identifier for a file retrievable from a [`SentrySourceConfig`].
#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
pub struct SentryFileId(String);

impl fmt::Display for SentryFileId {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Clone, Debug, serde::Deserialize)]
struct SearchResult {
    id: SentryFileId,
    // TODO: Add more fields
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
struct SearchQuery {
    index_url: Url,
    token: String,
}

/// An LRU cache sentry DIF index responses.
type SentryIndexCache = lru::LruCache<SearchQuery, (Instant, Vec<SearchResult>)>;

/// Errors happening while fetching data from Sentry.
#[derive(Debug, Error)]
pub enum SentryError {
    #[error("failed sending request to Sentry")]
    Reqwest(#[from] reqwest::Error),

    #[error("bad status code from Sentry: {0}")]
    BadStatusCode(StatusCode),

    #[error("failed joining task")]
    JoinTask(#[from] tokio::task::JoinError),
}

pub struct SentryDownloader {
    client: reqwest::Client,
    runtime: tokio::runtime::Handle,
    index_cache: Mutex<SentryIndexCache>,
    cache_duration: Duration,
    connect_timeout: Duration,
    streaming_timeout: Duration,
}

impl fmt::Debug for SentryDownloader {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct(std::any::type_name::<Self>())
            .field("connector", &format_args!("Addr(ClientConnector)"))
            .field("index_cache", &self.index_cache)
            .finish()
    }
}

impl SentryDownloader {
    pub fn new(client: reqwest::Client, runtime: tokio::runtime::Handle, config: &Config) -> Self {
        // The Sentry cache index should expire as soon as we attempt to retry negative caches.
        let cache_duration = if config.cache_dir.is_some() {
            config
                .caches
                .downloaded
                .retry_misses_after
                .unwrap_or_else(|| Duration::from_secs(0))
        } else {
            Duration::from_secs(0)
        };
        Self {
            client,
            runtime,
            index_cache: Mutex::new(SentryIndexCache::new(100_000)),
            cache_duration,
            connect_timeout: config.connect_timeout,
            streaming_timeout: config.streaming_timeout,
        }
    }

    /// Make a request to sentry, parse the result as a JSON SearchResult list.
    async fn fetch_sentry_json(
        client: &reqwest::Client,
        query: &SearchQuery,
    ) -> Result<Vec<SearchResult>, SentryError> {
        let mut request = client
            .get(query.index_url.clone())
            .bearer_auth(&query.token)
            .header("Accept-Encoding", "identity")
            .header("User-Agent", USER_AGENT);
        if let Some(span) = sentry::configure_scope(|scope| scope.get_span()) {
            for (k, v) in span.iter_headers() {
                request = request.header(k, v);
            }
        }

        let response = request.send().await?;

        if response.status().is_success() {
            tracing::trace!("Success fetching index from Sentry");
            Ok(response.json().await?)
        } else {
            tracing::warn!("Sentry returned status code {}", response.status());
            Err(SentryError::BadStatusCode(response.status()))
        }
    }

    /// Return the search results.
    ///
    /// If there are cached search results this skips the actual search.
    async fn cached_sentry_search(
        &self,
        query: SearchQuery,
    ) -> Result<Vec<SearchResult>, DownloadError> {
        if let Some((created, entries)) = self.index_cache.lock().get(&query) {
            if created.elapsed() < self.cache_duration {
                return Ok(entries.clone());
            }
        }

        tracing::debug!(
            "Fetching list of Sentry debug files from {}",
            &query.index_url
        );

        let entries = {
            let client = self.client.clone();
            let query = query.clone();
            let future = async move {
                future_utils::retry(|| Self::fetch_sentry_json(&client, &query)).await
            };

            let future =
                CancelOnDrop::new(self.runtime.spawn(future.bind_hub(sentry::Hub::current())));

            future.await.map_err(SentryError::JoinTask)??
        };

        if self.cache_duration > Duration::from_secs(0) {
            self.index_cache
                .lock()
                .put(query, (Instant::now(), entries.clone()));
        }

        Ok(entries)
    }

    pub async fn list_files(
        &self,
        source: Arc<SentrySourceConfig>,
        object_id: ObjectId,
        file_types: &[FileType],
    ) -> Result<Vec<RemoteDif>, DownloadError> {
        // TODO(flub): These queries do not handle pagination.  But sentry only starts to
        // paginate at 20 results so we get away with this for now.

        // There needs to be either a debug_id or a code_id filter in the query. Otherwise, this would
        // return a list of all debug files in the project.
        if object_id.debug_id.is_none() && object_id.code_id.is_none() {
            return Ok(Vec::new());
        }

        let mut index_url = source.url.clone();
        if let Some(ref debug_id) = object_id.debug_id {
            index_url
                .query_pairs_mut()
                .append_pair("debug_id", &debug_id.to_string());
        }

        // See <sentry-repo>/src/sentry/constants.py KNOWN_DIF_FORMATS for these query strings.
        index_url.query_pairs_mut().extend_pairs(
            file_types
                .iter()
                .map(|file_type| match file_type {
                    FileType::UuidMap => "uuidmap",
                    FileType::BcSymbolMap => "bcsymbolmap",
                    FileType::Pe => "pe",
                    FileType::Pdb => "pdb",
                    FileType::MachDebug | FileType::MachCode => "macho",
                    FileType::ElfDebug | FileType::ElfCode => "elf",
                    FileType::WasmDebug | FileType::WasmCode => "wasm",
                    FileType::Breakpad => "breakpad",
                    FileType::SourceBundle => "sourcebundle",
                    FileType::Il2cpp => "il2cpp",
                })
                .map(|val| ("file_formats", val)),
        );

        if let Some(ref code_id) = object_id.code_id {
            index_url
                .query_pairs_mut()
                .append_pair("code_id", code_id.as_str());
        }

        let query = SearchQuery {
            index_url,
            token: source.token.clone(),
        };

        let search = self.cached_sentry_search(query).await?;
        let file_ids = search
            .into_iter()
            .map(|search_result| SentryRemoteDif::new(source.clone(), search_result.id).into())
            .collect();

        Ok(file_ids)
    }

    /// Downloads a source hosted on Sentry.
    ///
    /// # Directly thrown errors
    /// - [`DownloadError::Reqwest`]
    /// - [`DownloadError::Rejected`]
    /// - [`DownloadError::Canceled`]
    pub async fn download_source(
        &self,
        file_source: SentryRemoteDif,
        destination: &Path,
    ) -> Result<DownloadStatus, DownloadError> {
        let request = self
            .client
            .get(file_source.url())
            .header("User-Agent", USER_AGENT)
            .bearer_auth(&file_source.source.token)
            .send();

        let download_url = file_source.url();
        let source = RemoteDif::from(file_source);
        let request = tokio::time::timeout(self.connect_timeout, request);
        let request = super::measure_download_time(source.source_metric_key(), request);

        match request.await {
            Ok(Ok(response)) => {
                if response.status().is_success() {
                    tracing::trace!("Success hitting {}", download_url);

                    let content_length = response
                        .headers()
                        .get(header::CONTENT_LENGTH)
                        .and_then(|hv| hv.to_str().ok())
                        .and_then(|s| s.parse::<u32>().ok());

                    let timeout =
                        content_length.map(|cl| content_length_timeout(cl, self.streaming_timeout));
                    let stream = response.bytes_stream().map_err(DownloadError::Reqwest);

                    super::download_stream(&source, stream, destination, timeout).await
                } else if matches!(
                    response.status(),
                    StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED
                ) {
                    tracing::debug!("Insufficient permissions to download from {}", download_url);
                    Err(DownloadError::Permissions)
                } else if response.status().is_client_error() {
                    tracing::debug!(
                        "Unexpected client error status code from {}: {}",
                        download_url,
                        response.status()
                    );
                    Ok(DownloadStatus::NotFound)
                } else {
                    tracing::debug!(
                        "Unexpected status code from {}: {}",
                        download_url,
                        response.status()
                    );
                    Err(DownloadError::Rejected(response.status()))
                }
            }
            Ok(Err(e)) => {
                tracing::debug!("Skipping response from {}: {}", download_url, e);
                Err(DownloadError::Reqwest(e)) // must be wrong type
            }
            // Timed out
            Err(_) => Err(DownloadError::Canceled),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::sources::SourceId;

    #[test]
    fn test_download_url() {
        let source = SentrySourceConfig {
            id: SourceId::new("test"),
            url: Url::parse("https://example.net/endpoint/").unwrap(),
            token: "token".into(),
        };
        let file_source = SentryRemoteDif::new(Arc::new(source), SentryFileId("abc123".into()));
        let url = file_source.url();
        assert_eq!(url.as_str(), "https://example.net/endpoint/?id=abc123");
    }

    #[test]
    fn test_uri() {
        let source = SentrySourceConfig {
            id: SourceId::new("test"),
            url: Url::parse("https://example.net/endpoint/").unwrap(),
            token: "token".into(),
        };
        let file_source = SentryRemoteDif::new(Arc::new(source), SentryFileId("abc123".into()));
        let uri = file_source.uri();
        assert_eq!(uri, RemoteDifUri::new("sentry://project_debug_file/abc123"));
    }
}
