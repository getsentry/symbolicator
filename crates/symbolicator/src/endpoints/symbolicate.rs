use axum::extract;
use axum::response::Json;
use serde::{Deserialize, Serialize};

use symbolicator_sources::SourceConfig;

use crate::service::{
    RawObjectInfo, RawStacktrace, RequestOptions, RequestService, Scope, Signal, StacktraceOrigin,
    SymbolicateStacktraces, SymbolicationResponse,
};
use crate::utils::sentry::ConfigureScope;

use super::ResponseError;

/// Query parameters of the symbolication request.
#[derive(Deserialize)]
pub struct SymbolicationRequestQueryParams {
    #[serde(default)]
    pub timeout: Option<u64>,
    #[serde(default)]
    pub scope: Scope,
}

impl ConfigureScope for SymbolicationRequestQueryParams {
    fn to_scope(&self, scope: &mut sentry::Scope) {
        scope.set_tag("request.scope", &self.scope);
        if let Some(timeout) = self.timeout {
            scope.set_tag("request.timeout", timeout);
        } else {
            scope.set_tag("request.timeout", "none");
        }
    }
}

/// JSON body of the symbolication request.
#[derive(Serialize, Deserialize)]
pub struct SymbolicationRequestBody {
    #[serde(default)]
    pub signal: Option<Signal>,
    #[serde(default)]
    pub sources: Option<Vec<SourceConfig>>,
    #[serde(default)]
    pub stacktraces: Vec<RawStacktrace>,
    #[serde(default)]
    pub modules: Vec<RawObjectInfo>,
    #[serde(default)]
    pub options: RequestOptions,
}

pub async fn symbolicate_frames(
    extract::State(service): extract::State<RequestService>,
    extract::Query(params): extract::Query<SymbolicationRequestQueryParams>,
    extract::Json(body): extract::Json<SymbolicationRequestBody>,
) -> Result<Json<SymbolicationResponse>, ResponseError> {
    sentry::start_session();

    params.configure_scope();

    let sources = match body.sources {
        Some(sources) => sources.into(),
        None => service.config().default_sources(),
    };

    let request_id = service.symbolicate_stacktraces(
        SymbolicateStacktraces {
            scope: params.scope,
            signal: body.signal,
            sources,
            origin: StacktraceOrigin::Symbolicate,
            stacktraces: body.stacktraces,
            modules: body.modules.into_iter().map(From::from).collect(),
        },
        body.options,
    )?;

    match service.get_response(request_id, params.timeout).await {
        Some(response) => Ok(Json(response)),
        None => Err("symbolication request did not start".into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use reqwest::{Client, StatusCode};
    use symbolicator_service::types::CompletedSymbolicationResponse;

    use crate::test;

    #[tokio::test]
    async fn test_body_limit() {
        test::setup();

        let server = test::server_with_default_service().await;

        let mut buf = vec![b'.'; 4 * 1024 * 1024];
        buf[0] = b'"';
        *buf.last_mut().unwrap() = b'"';

        let response = Client::new()
            .post(server.url("/symbolicate"))
            .header("Content-Type", "application/json")
            .body(buf)
            .send()
            .await
            .unwrap();

        // the JSON does not fit our schema :-)
        assert_eq!(response.status(), StatusCode::UNPROCESSABLE_ENTITY);

        let mut buf = vec![b'.'; 8 * 1024 * 1024];
        buf[0] = b'"';
        *buf.last_mut().unwrap() = b'"';

        let response = Client::new()
            .post(server.url("/symbolicate"))
            .header("Content-Type", "application/json")
            .body(buf)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::PAYLOAD_TOO_LARGE);
    }

    #[tokio::test]
    async fn test_unknown_field() {
        test::setup();

        let server = test::server_with_default_service().await;

        let payload = r##"{
            "stacktraces": [],
            "modules": [],
            "sources": [],
            "unknown": "value"
        }"##;

        let response = Client::new()
            .post(server.url("/symbolicate"))
            .header("Content-Type", "application/json")
            .body(payload)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }

    /// Asserts that disabling requesting for DIF candidates info works.
    #[tokio::test]
    async fn test_no_dif_candidates() {
        test::setup();

        let server = test::server_with_default_service().await;

        let payload = r##"{
            "stacktraces": [{
              "registers": {"eip": "0x0000000001509530"},
              "frames": [{"instruction_addr": "0x749e8630"}]
            }],
            "modules": [{
              "type": "pe",
              "debug_id": "ff9f9f78-41db-88f0-cded-a9e1e9bff3b5-1",
              "code_file": "C:\\Windows\\System32\\kernel32.dll",
              "debug_file": "C:\\Windows\\System32\\wkernel32.pdb",
              "image_addr": "0x749d0000",
              "image_size": 851968
            }],
            "sources": []
        }"##;
        let mut payload: SymbolicationRequestBody = serde_json::from_str(payload).unwrap();
        payload.sources = Some(vec![test::microsoft_symsrv()]);

        let response = Client::new()
            .post(server.url("/symbolicate"))
            .json(&payload)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        // TODO: assert symbol server hits as side-effect?

        let response: CompletedSymbolicationResponse = response.json().await.unwrap();
        test::assert_snapshot!(response);
    }

    /// Requests could contain invalid data which should not stop symbolication,
    /// name is sent by Sentry and is unknown.
    #[tokio::test]
    async fn test_unknown_source_config() {
        test::setup();

        let server = test::server_with_default_service().await;

        let payload = r##"{
            "stacktraces": [{
              "registers": {"eip": "0x0000000001509530"},
              "frames": [{"instruction_addr": "0x749e8630"}]
            }],
            "modules": [{
              "type": "pe",
              "debug_id": "ff9f9f78-41db-88f0-cded-a9e1e9bff3b5-1",
              "code_file": "C:\\Windows\\System32\\kernel32.dll",
              "debug_file": "C:\\Windows\\System32\\wkernel32.pdb",
              "image_addr": "0x749d0000",
              "image_size": 851968
            }],
            "sources": [{
              "type": "http",
              "id": "unknown",
              "layout": {"type": "symstore"},
              "filters": {"filetypes": ["mach_code"]},
              "url": "http://notfound",
              "name": "not a known field",
              "not-a-field": "more unknown fields"
            }],
            "options": {"dif_candidates": true}
        }"##;

        let response = Client::new()
            .post(server.url("/symbolicate"))
            .header("Content-Type", "application/json")
            .body(payload)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let response: CompletedSymbolicationResponse = response.json().await.unwrap();
        test::assert_snapshot!(response);
    }
}
