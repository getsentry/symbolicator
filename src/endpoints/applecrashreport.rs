use actix_web::{error, multipart, App, Error, HttpMessage, HttpRequest, Json, Query, State};
use futures::{compat::Stream01CompatExt, StreamExt};

use crate::endpoints::symbolicate::SymbolicationRequestQueryParams;
use crate::services::Service;
use crate::types::{RequestOptions, SymbolicationResponse};
use crate::utils::multipart::{
    read_multipart_file, read_multipart_request_options, read_multipart_sources,
};
use crate::utils::sentry::ConfigureScope;

async fn handle_apple_crash_report_request(
    state: State<Service>,
    params: Query<SymbolicationRequestQueryParams>,
    request: HttpRequest<Service>,
) -> Result<Json<SymbolicationResponse>, Error> {
    sentry::start_session();

    let params = params.into_inner();
    params.configure_scope();

    let mut report = None;
    let mut sources = state.config().default_sources();
    let mut options = RequestOptions::default();

    let mut stream = request.multipart().compat();
    while let Some(item) = stream.next().await {
        let field = match item? {
            multipart::MultipartItem::Field(field) => field,
            _ => return Err(error::ErrorBadRequest("unsupported nested formdata")),
        };

        let content_disposition = field.content_disposition();
        match content_disposition.as_ref().and_then(|d| d.get_name()) {
            Some("apple_crash_report") => report = Some(read_multipart_file(field).await?),
            Some("sources") => sources = read_multipart_sources(field).await?.into(),
            Some("options") => options = read_multipart_request_options(field).await?,
            _ => (), // Always ignore unknown fields.
        }
    }

    let report = report.ok_or_else(|| error::ErrorBadRequest("missing apple crash report"))?;

    let symbolication = state.symbolication();
    let request_id =
        symbolication.process_apple_crash_report(params.scope, report, sources, options);

    match symbolication.get_response(request_id, params.timeout).await {
        Some(response) => Ok(Json(response)),
        None => Err(error::ErrorInternalServerError(
            "symbolication request did not start",
        )),
    }
}

pub fn configure(app: App<Service>) -> App<Service> {
    app.resource("/applecrashreport", |r| {
        let handler = compat_handler!(handle_apple_crash_report_request, s, p, r);
        r.post().with_async(handler);
    })
}

#[cfg(test)]
mod tests {
    use std::fs;

    use actix_web::test::TestServer;
    use reqwest::{multipart, Client, StatusCode};

    use crate::config::Config;
    use crate::services::Service;
    use crate::test;
    use crate::types::SymbolicationResponse;

    #[tokio::test]
    async fn test_basic() {
        test::setup();

        let service = Service::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));

        let file_contents = fs::read("tests/fixtures/apple_crash_report.txt").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("apple_crash_report.txt");

        let form = multipart::Form::new()
            .part("apple_crash_report", file_part)
            .text("sources", "[]");

        let response = Client::new()
            .post(&server.url("/applecrashreport"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let body = response.text().await.unwrap();
        let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
        insta::assert_yaml_snapshot!(response);
    }

    #[tokio::test]
    async fn test_unknown_field() {
        test::setup();

        let service = Service::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));

        let file_contents = fs::read("tests/fixtures/apple_crash_report.txt").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("apple_crash_report.txt");

        let form = multipart::Form::new()
            .part("apple_crash_report", file_part)
            .text("sources", "[]")
            .text("unknown", "value");

        let response = Client::new()
            .post(&server.url("/applecrashreport"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }
}
