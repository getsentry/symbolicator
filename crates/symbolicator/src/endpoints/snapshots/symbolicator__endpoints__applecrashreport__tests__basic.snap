---
source: src/endpoints/applecrashreport.rs
expression: response
---
status: completed
timestamp: 1547055742
system_info:
  os_name: macOS
  os_version: 10.14.0
  os_build: 18A391
  cpu_arch: x86_64
  device_model: "MacBookPro14,3"
crashed: true
crash_reason: SIGSEGV
crash_details: "objc_msgSend() selector name: respondsToSelector:\n  more information here"
stacktraces:
  - thread_id: 0
    is_requesting: false
    frames:
      - status: unknown_image
        original_index: 0
        instruction_addr: "0x7fff61bc6c2a"
        package: libsystem_kernel.dylib
      - status: unknown_image
        original_index: 1
        instruction_addr: "0x7fff349f505e"
        package: CoreFoundation
      - status: unknown_image
        original_index: 2
        instruction_addr: "0x7fff349f45ad"
        package: CoreFoundation
      - status: unknown_image
        original_index: 3
        instruction_addr: "0x7fff349f3ce4"
        package: CoreFoundation
      - status: unknown_image
        original_index: 4
        instruction_addr: "0x7fff33c8d895"
        package: HIToolbox
      - status: unknown_image
        original_index: 5
        instruction_addr: "0x7fff33c8d5cb"
        package: HIToolbox
      - status: unknown_image
        original_index: 6
        instruction_addr: "0x7fff33c8d348"
        package: HIToolbox
      - status: unknown_image
        original_index: 7
        instruction_addr: "0x7fff31f4a95b"
        package: AppKit
      - status: unknown_image
        original_index: 8
        instruction_addr: "0x7fff31f496fa"
        package: AppKit
      - status: unknown_image
        original_index: 9
        instruction_addr: "0x7fff31f4375d"
        package: AppKit
      - status: missing
        original_index: 10
        instruction_addr: "0x108b7092b"
        package: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/MacOS/YetAnotherMac
      - status: missing
        original_index: 11
        instruction_addr: "0x108b702a6"
        package: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/MacOS/YetAnotherMac
      - status: unknown_image
        original_index: 12
        instruction_addr: "0x7fff61a8e085"
        package: libdyld.dylib
      - status: unknown_image
        original_index: 13
        instruction_addr: "0xea004"
        package: YetanotherMac
  - thread_id: 1
    is_requesting: true
    registers:
      cs: "0x2b"
      fs: "0x0"
      gs: "0x0"
      r10: "0x0"
      r11: "0xffffffff"
      r12: "0x8"
      r13: "0x11e800b00"
      r14: "0x1"
      r15: "0x0"
      r8: "0x3"
      r9: "0x10"
      rax: "0x20261bb4775b008f"
      rbp: "0x700015a616d0"
      rbx: "0x0"
      rcx: "0x1288266c0"
      rdi: "0x0"
      rdx: "0x1"
      rflags: "0x10206"
      rip: "0x1090a0132"
      rsi: "0x0"
      rsp: "0x700015a613f0"
    frames:
      - status: unknown_image
        original_index: 0
        instruction_addr: "0x7fff61bc85be"
        package: libsystem_kernel.dylib
      - status: unknown_image
        original_index: 1
        instruction_addr: "0x7fff61c7f415"
        package: libsystem_pthread.dylib
      - status: unknown_image
        original_index: 2
        instruction_addr: "0x54485244"
modules:
  - debug_status: missing
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: macho
    code_id: 2d903291397d3d14bfca52c7fb8c5e00
    code_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/MacOS/YetAnotherMac
    debug_id: 2d903291-397d-3d14-bfca-52c7fb8c5e00
    debug_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/MacOS/YetAnotherMac
    image_addr: "0x10864e000"
    image_size: 108797951
  - debug_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: macho
    code_id: 6deccee4a0523ea4bb67957b06f53ad1
    code_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPhysX3PROFILE.dylib
    debug_id: 6deccee4-a052-3ea4-bb67-957b06f53ad1
    debug_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPhysX3PROFILE.dylib
    image_addr: "0x112bb2000"
    image_size: 2170879
  - debug_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: macho
    code_id: 5e012a646cc536f19b4da0564049169b
    code_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPhysX3CookingPROFILE.dylib
    debug_id: 5e012a64-6cc5-36f1-9b4d-a0564049169b
    debug_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPhysX3CookingPROFILE.dylib
    image_addr: "0x112fc0000"
    image_size: 221183
  - debug_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: macho
    code_id: 9c19854471943de6b67e4cc27eed2eab
    code_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPhysX3CommonPROFILE.dylib
    debug_id: 9c198544-7194-3de6-b67e-4cc27eed2eab
    debug_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPhysX3CommonPROFILE.dylib
    image_addr: "0x113013000"
    image_size: 1474559
  - debug_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: macho
    code_id: 890f0997f90435449af7cf011f09a06e
    code_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPxFoundationPROFILE.dylib
    debug_id: 890f0997-f904-3544-9af7-cf011f09a06e
    debug_file: /Users/bruno/Documents/Unreal Projects/YetAnotherMac/MacNoEditor/YetAnotherMac.app/Contents/UE4/Engine/Binaries/ThirdParty/PhysX3/Mac/libPxFoundationPROFILE.dylib
    image_addr: "0x1131fa000"
    image_size: 28671
